source/Calamari.Aws/Kubernetes/Discovery/AwsKubernetesDiscoverer.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon;
+using Amazon.EKS;
+using Amazon.EKS.Model;
+using Calamari.Common.Features.Discovery;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Kubernetes.Aws;
+using Newtonsoft.Json;
+using Octopus.CoreUtilities.Extensions;
+
+namespace Calamari.Aws.Kubernetes.Discovery
+{
+    public class AwsKubernetesDiscoverer : IKubernetesDiscoverer
+    {
+        readonly ILog log;
+
+        public AwsKubernetesDiscoverer(ILog log)
+        {
+            this.log = log;
+        }
+
+        public string Name => "Aws";
+        
+        public IEnumerable<KubernetesCluster> DiscoverClusters(string contextJson, IVariables variables)
+        {
+            if (!TryGetAuthenticationDetails(contextJson, out var authenticationDetails))
+                yield break;
+            var workerPool = variables.Get("Octopus.Aws.WorkerPool");
+            
+            var accessKeyOrWorkerCredentials = authenticationDetails.Credentials.Type == "account"
+                ? $"Access Key: {authenticationDetails.Credentials.Account.AccessKey}"
+                : $"Using Worker Credentials on Worker Pool: {workerPool}";
+
+            log.Verbose("Looking for Kubernetes clusters in AWS using:");
+            log.Verbose($"  Regions: [{string.Join(",",authenticationDetails.Regions)}]");
+            
+            log.Verbose("  Account:");
+            log.Verbose($"    {accessKeyOrWorkerCredentials}");
+            
+            if (authenticationDetails.Role.Type == "assumeRole")
+            {
+                log.Verbose("  Role:");
+                log.Verbose($"    ARN: {authenticationDetails.Role.Arn}");
+                if (!authenticationDetails.Role.SessionName.IsNullOrEmpty())
+                    log.Verbose($"    Session Name: {authenticationDetails.Role.SessionName}");
+                if (authenticationDetails.Role.SessionDuration != null) 
+                    log.Verbose($"    Session Duration: {authenticationDetails.Role.SessionDuration}");
+                if (!authenticationDetails.Role.ExternalId.IsNullOrEmpty())
+                    log.Verbose($"    External Id: {authenticationDetails.Role.ExternalId}");
+            }
+            else
+            {
+                log.Verbose("  Role: No IAM Role provided.");
+            }
+            
+            foreach (var region in authenticationDetails.Regions)
+            {
+                var client = new AmazonEKSClient(authenticationDetails.ToCredentials(),
+                    RegionEndpoint.GetBySystemName(region));
+
+                var clusters = client.ListClustersAsync(new ListClustersRequest()).GetAwaiter().GetResult();
+                
+                log.Info("Discovery has discovered");
```

Comment:
Is this logging supposed to be here? If so, it probably needs rewording.

Reply:
Oops, good spot!

Reply:
I've removed this because it isn't required. We log how many clusters have been discovered in `KubernetesDiscoveryCommand.cs`